Implementation of IndexIVFRaBitQFastScan

alibeklfc opened the pull request:
Summary:
**Introduction**

This diff implements `IndexIVFRaBitQFastScan`, a high-performance approximate nearest neighbor search index that combines Inverted File (IVF) partitioning, RaBitQ quantization, and FastScan batch processing for significant performance improvements in vector similarity search.

**Implementation**
* **New Source and Header Files**: Added implementations for IndexIVFRaBitQFastScan, following a similar interface to IndexIVFRaBitQ.

* **Batched Processing**: The new index inherits from `IndexIVFFastScan`. It processes 32 database vectors simultaneously using SIMD operations.

* **Specialized Post-processing Handler**:  Extends the FastScan framework with RaBitQ-aware handler factories that automatically detect and optimize for RaBitQ indices, maintaining compatibility with existing IVF search parameters while enabling specialized optimizations.

* **LUT**: IndexRaBitQFastScan produces slightly different results than IndexIVFRaBitQ due to an extra quantization step in the IndexFastScan architecture. Computes lookup tables using query residuals (query minus coarse centroid) with probe-specific query factors, ensuring accurate distance calculations across different inverted list probes.
* **Centered parameter**: Implements both centered and non-centered query quantization modes

* **Query Offset Parameter**: Added a query offset parameter similarly as in D81787307

**Testing**
* Includes extensive test suite covering performance benchmarks, distance calculation correctness, recall measurements, and cross-validation against reference implementations

* All tests passed successfully, validating the correctness and robustness of IndexRaBitQFastScan.

**Results**
 ivfrabitq2
 ivfrabitq1
* **Performance Dependency**: Performance measurements confirm that IndexIVFRaBitQFastScan is notably faster than IndexIVFRaBitQ when the qb value is high. While the original IndexIVFRaBitQ experiences increased runtime with higher qb values, the fast scan variant maintains consistent runtime regardless of qb.

Differential Revision: D83264335

alexanderguzhva commented on faiss/IndexFastScan.cpp:
```diff
@@ -163,11 +164,23 @@ void estimators_from_tables_generic(
     using accu_t = typename C::T;
 
+    // Check if this is RaBitQ for specialized distance computation
+    const IndexRaBitQFastScan* rabitq_index =
```
This totally violates the fundamentals of object-oriented programming. The base class `IndexFastScan` must know nothing about a derived class `IndexRaBitQFastScan`. Please consider reorganizing this.

alexanderguzhva commented on faiss/IndexFastScan.cpp:
```diff
@@ -454,25 +539,55 @@ void IndexFastScan::search_implem_12(
+    if (rabitq_index != nullptr) {
+        // Use RaBitQHeapHandler for RaBitQ (direct processing, no
+        // post-processing needed)
+        RaBitQHeapHandler<C> rabitq_handler(
```
same story here and below

alexanderguzhva commented on faiss/IndexIVFFastScan.cpp:
```diff
@@ -400,7 +402,29 @@ SIMDResultHandlerToFloat* make_knn_handler(
+        idx_t query_offset = 0) {
+    // Check if this is IndexIVFRaBitQFastScan and create specialized handler
+    if (index) {
+        auto* rabitq_index = dynamic_cast<const IndexIVFRaBitQFastScan*>(index);
```
same story about `IndexIVFFastScan` and `IndexIVFRaBitQFastScan`

alexanderguzhva commented on faiss/IndexIVFRaBitQFastScan.cpp:
```diff
@@ -0,0 +1,694 @@
+    float v_min = std::numeric_limits<float>::max();
+    float v_max = std::numeric_limits<float>::lowest();
+
+    if (centered) {
```
do I get it correct that this is a copy-pasted code from `impl/RaBitQuantizer.cpp`? This creates a need to synchronize the changes between this code and `impl/RaBitQuantizer.cpp`. Please reorganize the code to have only 1 instance of these facilities, if possible

alexanderguzhva commented:
@alibeklfc I would strongly recommend refactoring the candidate code

alibeklfc commented:
@alexanderguzhva Thank you for your valuable feedback!

I have addressed your comments. The IndexFastScan and IndexIVFFastScan do not know anything about the child classes now. I have made the make_knn_handler function part of the main class and we override it in the child classes.

I have also created the RaBitQUtils namespace to eliminate code duplication.

alexanderguzhva commented:
@alibeklfc thanks, let me go through both diffs carefully. Would you like to have #4595 commited prior to #4596?

alexanderguzhva commented on faiss/IndexIVFFastScan.cpp:
```diff
@@ -971,6 +994,11 @@ void IndexIVFFastScan::search_implem_10(
             handler.id_map = ids.get();
 
+            // Set context information for handlers that need additional data
+            std::vector<int> probe_map(1);
```
this allocation operation is a costly one. Would it be possible to allocate this vector once, before `for (idx_t i = 0; i < n; i++) {` line?

Also, if you need to pass vectors (like down below), then the best strategy is the following:
```C++
std::vector<int> some_vector;
some_vector.reserve(initial_guess);

for (size_t i = 0; i < n; i++) {
    ...
    // the trick is that an internal vector implementation may skip immediately shrinking
    //   the amount of allocated memory, so this operation is very cheap, especially if
    //   your `initial_guess` size covers all new sizes  
    some_vector.resize(new_size);

    // use the vector
    ...
}
```
This allows you not to waste your time on memory allocations

alexanderguzhva commented on faiss/IndexIVFRaBitQFastScan.h:
```diff
@@ -0,0 +1,212 @@
+
+    /// Extracted factors storage for batch processing
+    /// Size: ntotal, stores factors separately from packed codes
+    std::vector<FactorsData> factors_storage;
```
this needs to be changed, because the current scenario makes the it thread-unsafe.
Please consider the following possible pattern:
```C++
struct SAPreprocessed {};
struct IndexIVFFastScan {
    virtual void sa_encode_preprocess(
            idx_t n,
            const uint8_t* flat_codes,
            idx_t start_global_idx,
            SAPreprocessed* sa_preprocessed
    ) = {};

    virtual void IndexIVFFastScan::consume_preprocessed(
            const SAPreprocessed* preprocessed
    ) = {};
}; 

struct RaBitQSAPreprocessed: public SAPreprocessed {
    std::vector<float> factors;
};

void IndexIVFRaBitQFastScan::sa_encode_preprocess(..., SAPreprocessed* preprocessed) {
    RaBitQSAPreprocessed* rabitq_preprocessed = dynamic_cast<RaBitQSAPreprocessed*>(preprocessed);
    // populate rabitq_preprocessed
    ...
}

void IndexIVFRaBitQFastScan::consume_preprocessed(..., const SAPreprocessed* preprocessed) {
    RaBitQSAPreprocessed* rabitq_preprocessed = dynamic_cast<RaBitQSAPreprocessed*>(preprocessed);
    // use rabitq_preprocessed
}
```
